### ISSUE TITLE
ibm_db connect() blocks main thread when server not responding until timeout

### ISSUE THREAD

**Issue by kotofos** [ibmdb/python-ibmdb#703]
Hi, while working on SlashDB, we faced an issue that the .connect() function in the thread blocks all the threads while it trying to connect. 
When a server is down, or IP is wrong the connection will block all threads until it times out. So the whole web server hangs. To reproduce firewall must drop the packages, so it will timeout, not just quick reject.

Seems that only .connect() is affected. The .execute(), even long, does not block. 
Please see attached code to reproduce the issue.

Tested on ibm_db installed by pip. 
* 3.0.3 
* 3.0.4 
* 3.1.0 
* 2.0.5.1

The cli driver is handled by pip too.
The platform is Linux.

```
from threading import Thread
from time import sleep
import ibm_db
import ibm_db_dbi
from sqlalchemy import create_engine

# blocks main
def sa_connect():
    print('connecting')
    engine = create_engine('db2://a:a@1.2.3.4')
    engine.connect()
    print('done')

# blocks main
def ibm_db_connect():
    print('connecting')
    ibm_db.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
    print('done')

# blocks main
def ibm_dbi_connect():
    print('connecting')
    ibm_db_dbi.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
    print('done')

# not blocks main
def sock_connect():
    import socket
    sock = socket.socket()
    sock.connect(('1.2.3.4', 60000))

# not blocks main
def delay_query():
    print('connecting')
    engine = create_engine('db2://user:pswd@127.0.0.1:50000/Chinook')
    engine.connect()
    print('executing delay')
    engine.execute('call dbms_alert.sleep(20)')

print('start')
thread = Thread(target=ibm_db_connect)
thread.start()
for i in range(20):
    print('.', end='')  # will stop printing until unblocked
    sleep(1)
    
thread.join()
print('end')
```

**Comment by amukherjee28** [comment]
@kotofos 

I did allow multiple threads to run along with your sample program and make mistake in 2 connection strings.

The one with Wrong Connection info did show me the error and other went through fine.

Here's the output

```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Users\amukherjee\python391\lib\threading.py", line 954, in _bootstrap_inner
    self.run()
  File "C:\Users\amukherjee\python391\lib\threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\amukherjee\Working_repository\open_source_work\python_ibm_db\samples\thread5.py", line 17, in ibm_db_connect
    ibm_db.connect("DATABASE=BJHAD;HOSTNAME=HOSTNAME;PORT=12345;PROTOCOL=TCPIP;UID=USERID;PWD=PASSWORD;", "", "")
Exception: [IBM][CLI Driver] SQL30061N  The database alias or database name "BJHAD             " was not found at the remote node.  SQLSTATE=08004
 SQLCODE=-30061
Exception in thread Thread-2:
Traceback (most recent call last):
  File "C:\Users\amukherjee\python391\lib\site-packages\ibm_db_dbi.py", line 626, in connect
    conn = ibm_db.connect(dsn, '', '', conn_options)
Exception: [IBM][CLI Driver] SQL30061N  The database alias or database name "SAMPL             " was not found at the remote node.  SQLSTATE=08004
 SQLCODE=-30061

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 3212, in _wrap_pool_connect
    return fn()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 307, in connect
    return _ConnectionFairy._checkout(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 767, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 425, in checkout
    rec = pool._do_get()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\impl.py", line 146, in _do_get
    self._dec_overflow()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
    return self._create_connection()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 253, in _create_connection
    return _ConnectionRecord(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 368, in __init__
    self.__connect()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 611, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 605, in __connect
    connection = pool._invoke_creator(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\create.py", line 578, in connect
    return dialect.connect(*cargs, **cparams)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\default.py", line 584, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "C:\Users\amukherjee\python391\lib\site-packages\ibm_db_dbi.py", line 631, in connect
    raise _get_exception(inst)
ibm_db_dbi.OperationalError: ibm_db_dbi::OperationalError: [IBM][CLI Driver] SQL30061N  The database alias or database name "SAMPL             " was not found at the remote node.  SQLSTATE=08004\r SQLCODE=-30061

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\amukherjee\python391\lib\threading.py", line 954, in _bootstrap_inner
    self.run()
  File "C:\Users\amukherjee\python391\lib\threading.py", line 892, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Users\amukherjee\Working_repository\open_source_work\python_ibm_db\samples\thread5.py", line 11, in sa_connect
    engine.connect()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 3166, in connect
    return self._connection_cls(self, close_with_result=close_with_result)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 96, in __init__
    else engine.raw_connection()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 3245, in raw_connection
    return self._wrap_pool_connect(self.pool.connect, _connection)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 3215, in _wrap_pool_connect
    Connection._handle_dbapi_exception_noconnection(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 2069, in _handle_dbapi_exception_noconnection
    util.raise_(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\base.py", line 3212, in _wrap_pool_connect
    return fn()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 307, in connect
    return _ConnectionFairy._checkout(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 767, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 425, in checkout
    rec = pool._do_get()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\impl.py", line 146, in _do_get
    self._dec_overflow()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\impl.py", line 143, in _do_get
    return self._create_connection()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 253, in _create_connection
    return _ConnectionRecord(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 368, in __init__
    self.__connect()
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 611, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\util\compat.py", line 207, in raise_
    raise exception
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\pool\base.py", line 605, in __connect
    connection = pool._invoke_creator(self)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\create.py", line 578, in connect
    return dialect.connect(*cargs, **cparams)
  File "C:\Users\amukherjee\python391\lib\site-packages\sqlalchemy\engine\default.py", line 584, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "C:\Users\amukherjee\python391\lib\site-packages\ibm_db_dbi.py", line 631, in connect
    raise _get_exception(inst)
sqlalchemy.exc.OperationalError: (ibm_db_dbi.OperationalError) ibm_db_dbi::OperationalError: [IBM][CLI Driver] SQL30061N  The database alias or database name "SAMPL             " was not found at the remote node.  SQLSTATE=08004\r SQLCODE=-30061
(Background on this error at: https://sqlalche.me/e/14/e3q8)
start
connecting using ibm_db_connect
....................connecting using alchemy2
....................connecting using ibm_db_dbi
....................connecting using sqlalchemy
....................done ibm_db_dbi
executing delay
end
```

In this case the once that had wrong info throws the errors the others ended properly.

I hope this is the expected behavior.

In your case I guess there was a hang seen when any of the connection failed to connect.
In such case can you try using CONNECTION TIMEOUT keyword in either of the following places :
1. Connection string
2. db2dsdriver.cfg file
3. db2cli.ini file

Documentation link
https://www.ibm.com/docs/en/db2/11.5?topic=keywords-connectiontimeout

This keyword will ensure that connection times out after waiting for certain interval of time and move forward.
Although I did not use the keyword while running the sample program you have send, but I guess since you are seeing the hang you can give it a try.

Let me know your observation on the same.

**Comment by kotofos** [comment]
Hi, thanks, timeout settings helps to reduce the problem. Though even a short 2s timeout still blocks all threads.
To reproduce it should be not just a random error in the connection string, but some IP of non-existing server or a server configured to drop connections.
The `db2://a:a@1.2.3.4` from the example should be fine for this.

The current workaround is to do the connection check beforehand.

```
    sock = socket.socket()
    sock.settimeout(2)
    result = sock.connect_ex((host, port))
    if result != 0:
        raise

    engine = create_engine(connstr)
    engine.connect()
```
This code helps but is not 100% failproof. It still may block if sever is down just between the connection check and call to db2.connect()

I still believe in general this is unexpected and unwanted behavior, and it would be nice to have it fixed. Even by hacking the current workaround in the codebase.

**Comment by bimalkjha** [comment]
@kotofos Please share the test program to reproduce the problem as Eramman is unable to reproduce the issue. Thanks.

**Comment by kotofos** [comment]
The issue is not about not throwing or throwing the wrong errors from API, the issue is about stopping all the threads while the connection is not timeout.

Sample code
```
from threading import Thread
from time import sleep, time
import ibm_db


# blocks main
def ibm_db_connect():
    print('connecting')
    ibm_db.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
    print('thread done')


print('start')
thread = Thread(target=ibm_db_connect)
thread.start()
start_time = time()
for i in range(20):
    print('.', end='')  # will stop printing until unblocked
    sleep(0.5)
print('\nmain done')
print('main duration', time() - start_time, 'expected - about 10')

thread.join()
print('all end')

```

Actual result - printing of dots paused until timeout. Main thread is running for 140s instead of 10.
```
start
connecting
..Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/kotofos/.pyenv/versions/3.9.8/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/home/kotofos/.pyenv/versions/3.9.8/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/home/kotofos/.config/JetBrains/PyCharm2021.2/scratches/scratch.py", line 9, in ibm_db_connect
    ibm_db.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "TCP/IP".  Communication API being used: "SOCKETS".  Location where the error was detected: "1.2.3.4".  Communication function detecting the error: "connect".  Protocol specific error code(s): "110", "*", "*".  SQLSTATE=08001 SQLCODE=-30081
..................
main done
main duration 140.81787991523743 expected - about 10
all end
```

Expected - nonstop printing of dots.
```
start
connecting
....................
main done
main duration 10.011082887649536 expected - 10

Exception in thread Thread-1:
Traceback (most recent call last):
  File "/home/kotofos/.pyenv/versions/3.9.8/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/home/kotofos/.pyenv/versions/3.9.8/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/home/kotofos/.config/JetBrains/PyCharm2021.2/scratches/scratch.py", line 9, in ibm_db_connect
    ibm_db.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "TCP/IP".  Communication API being used: "SOCKETS".  Location where the error was detected: "1.2.3.4".  Communication function detecting the error: "connect".  Protocol specific error code(s): "110", "*", "*".  SQLSTATE=08001 SQLCODE=-30081
all end
```

**Comment by Earammak** [comment]
@kotofos  I have ran your sample script and found that there is no timeout issue for me.
Find the below log for your sample script [ Here i have given the wrong port info]

C:\Tasks>python3 script_703.py
start
connecting
Exception in thread Thread-1 (ibm_db_connect):
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Tasks\script_703.py", line 9, in ibm_db_connect
    ibm_db.connect("DATABASE=TESTDB;HOSTNAME=waldevdbclnxtst06.dev.rocketsoftware.com;PORT=60001;PROTOCOL=TCPIP;UID=username;PWD=Passowrd",'','')
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "TCP/IP".  Communication API being used: "SOCKETS".  Location where the error was detected: "10.117.19.22".  Communication function detecting the error: "recv".  Protocol specific error code(s): "10054", "*", "0".  SQLSTATE=08001  SQLCODE=-30081
....................
main done
main duration 10.45492672920227 expected - about 10
all end

C:\Tasks>

**Comment by kotofos** [comment]
In order to reproduce it must timeout. Please try to use same IP same connection string as in my sample.

It not probably it is Linux specific.

**Comment by Earammak** [comment]
@kotofos Can you please tell for what purpose you are using for that particular IP address. I'm not running in Linux actually.
Find the below output for the sample script:

C:\Tasks>python3 script_703.py
start
connecting
Exception in thread Thread-1 (ibm_db_connect):
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.1520.0_x64__qbz5n2kfra8p0\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\Tasks\script_703.py", line 9, in ibm_db_connect
    ibm_db.connect("DATABASE=TESTDB;HOSTNAME=1.2.3.4;PORT=60001;PROTOCOL=TCPIP;UID=newton;PWD=A2m8test",'','')
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "TCP/IP".  Communication API being used: "SOCKETS".  Location where the error was detected: "1.2.3.4".  Communication function detecting the error: "connect".  Protocol specific error code(s): "10060", "*", "*".  SQLSTATE=08001  SQLCODE=-30081
....................
main done
main duration 30.637592315673828 expected - about 10
all end

C:\Tasks>

This error is actually connection timeout error [Protocol specific error code(s): "10060].

**Comment by bimalkjha** [comment]
https://stackoverflow.com/questions/11968689/python-multithreading-wait-till-all-threads-finished

**Comment by kotofos** [comment]
We have user-configurable connection parameters, so any IP address should be handled as expected, without blocking all threads.

**Comment by imavo** [comment]
This reproduces easily on my linux (ubuntu 20.04) with the test program shown above.
Is this only a problem on linux?

Debugging shows that as soon as ibm_db.c calls (inside the child thread) the `SQLDriverConnectW(...)` then the main thread becomes blocked. 

The `ibm_db_connect_helper()` function does not bracket this `SQLDriverConnectW()` API with `Py_BEGIN_ALLOW_THREADS;` and `Py_END_ALLOW_THREADS;` although it does this for other cli APIs and I don't know the reason.

**Comment by agilevic** [comment]
@imavo  thanks for digging into this.

**Comment by imavo** [comment]
@Earammak  did you verify that the code changes on linux x64 using  the unmodified test script supplied by kotofos cause any difference in observed behaviour?  I ask this because when I build and install with the revised ibm_db code, the symptom is unchanged.

**Comment by Earammak** [comment]
@imavo Here is the below output for the latest script mentioned above.
> uname -a
Linux waldevdbcttst02 4.18.0-80.el8.x86_64 #1 SMP Wed Mar 13 12:02:46 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
> python3 test703.py
start
connecting
....................
main done
main duration 10.013392686843872 expected - about 10
Exception in thread Thread-1 (ibm_db_connect):
Traceback (most recent call last):
  File "/work/ek/python310/python_install/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "/work/ek/python310/python_install/lib/python3.10/threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "/work/ek/python_issue/test703.py", line 9, in ibm_db_connect
    ibm_db.connect("DATABASE=name;HOSTNAME=1.2.3.4;PORT=60000;PROTOCOL=TCPIP;UID=username;PWD=password;", "", "")
Exception: [IBM][CLI Driver] SQL30081N  A communication error has been detected. Communication protocol being used: "TCP/IP".  Communication API being used: "SOCKETS".  Location where the error was detected: "1.2.3.4".  Communication function detecting the error: "connect".  Protocol specific error code(s): "110", "*", "*".  SQLSTATE=08001 SQLCODE=-30081
all end

**Comment by imavo** [comment]
@Earammak That output shows that the problem is not fixed. I understand that the expection of the programmer is that *during the period* that the child-thread is timing out attempting to connect to a deliberately nonexistent IP-address THAT the main-thread should continue to process, and show some dots on the screen periodically to indicate that the main thread is  running concurrently and processing (and not blocked).  Instead, what you (and I) observe is that while the child-thread is timing-out inside clidriver on the connect(), the main thread appears to be not running and does not display any dots UNTIL AFTER the child thread fails to connect, which we understand to be an indicator that the main thread is somehow blocked by the action in the child thread, and gets unblocked when the connect() fails in the child thread.

**Comment by kotofos** [comment]
Hi, I tried the latest master on Linux, which is working as expected. I confirm that problem is fixed.

In the output `main duration 10.013392686843872 expected - about 10` mean the main duration is 10 seconds as it should be. Some confusion can be seen because of `end=''` in `print('.', end='')` is may buffered until linefeed is printed in some terminals.

When `end=''` is removed it works fine.

**Comment by imavo** [comment]
@kotofos , That is good news. I confirm that when removing the `end=''` that the behaviour is to show the dots on screen each on a new line while the other thread is timing out. It appears that the code changes to wrap Py_BEGIN_ALLOW_THREADS; and Py_END_ALLOW_THREADS ; around cli calls appears to have worked.

**Comment by bimalkjha** [comment]
The pull request #781 has delivered the fix of this issue as https://github.com/ibmdb/python-ibmdb/commit/a158da5f6bfe5b28edafef695b4a045413ba36af
Thanks for confirming that fix is working. Closing the issue now.
